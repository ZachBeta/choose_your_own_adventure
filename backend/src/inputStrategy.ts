// Strategy interface for dialog input
export interface InputStrategy {
  chooseDialogOption(options: any[]): Promise<number>;
}

export class UserInputStrategy implements InputStrategy {
  prompt: any;
  constructor(prompt: any) {
    this.prompt = prompt;
  }
  async chooseDialogOption(options: any[]): Promise<number> {
    const idx = Number(this.prompt('> choose: '));
    return idx;
  }
}

export class LuigiInputStrategy implements InputStrategy {
  async chooseDialogOption(options: any[]): Promise<number> {
    return 0; // Always select the first option
  }
}
