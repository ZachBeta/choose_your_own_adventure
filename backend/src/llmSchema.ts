import { z } from "zod";

export const SkillCheckSchema = z.object({
  part: z.string(),
  dc: z.number(),
});

export const DialogOptionSchema = z.object({
  id: z.string(),
  text: z.string(),
  skillCheck: SkillCheckSchema.optional(),
});

export const ThoughtCabinetSchema = z.object({
  part: z.string(),
  text: z.string(),
});

export const SceneResponseSchema = z.object({
  monologue: z.string().default("..."),
  thoughtCabinet: z.array(ThoughtCabinetSchema).default([]),
  dialog: z.array(DialogOptionSchema).default([]),
  skillCheckResult: z.any().optional(), // Refine as needed for your use case
});
