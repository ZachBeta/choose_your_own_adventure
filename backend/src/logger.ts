import fs from 'fs';
import path from 'path';

const logFilePath = path.join(__dirname, '../server.log');

function formatMessage(level: string, message: any, ...args: any[]): string {
  const timestamp = new Date().toISOString();
  let msg = '';
  if (typeof message === 'object') {
    msg = JSON.stringify(message, null, 2);
  } else {
    msg = String(message);
  }
  if (args.length) {
    msg += ' ' + args.map(arg => typeof arg === 'object' ? JSON.stringify(arg, null, 2) : String(arg)).join(' ');
  }
  return `[${timestamp}] [${level}] ${msg}`;
}

function writeLog(formatted: string, toConsole: boolean = false) {
  fs.appendFileSync(logFilePath, formatted + '\n', { encoding: 'utf8' });
  if (toConsole) {
    console.error(formatted);
  }
}

export function logInfo(message: any, ...args: any[]) {
  writeLog(formatMessage('INFO', message, ...args), false);
}

export function logError(message: any, ...args: any[]) {
  writeLog(formatMessage('ERROR', message, ...args), true);
}

export function logDebug(message: any, ...args: any[]) {
  writeLog(formatMessage('DEBUG', message, ...args), false);
}

