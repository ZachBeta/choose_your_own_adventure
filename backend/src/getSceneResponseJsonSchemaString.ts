import { SceneResponseSchema } from "./llmSchema";
import { zodToJsonSchema } from "zod-to-json-schema";

export function getSceneResponseJsonSchemaString() {
  const schemaObj = zodToJsonSchema(SceneResponseSchema, "SceneResponse");
  return JSON.stringify(schemaObj, null, 2);
}
