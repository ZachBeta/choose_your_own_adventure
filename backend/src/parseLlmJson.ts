import { SceneResponseSchema } from "./llmSchema";

/**
 * Strips code fences and tries to extract JSON from LLM output, then validates with Zod.
 */
export function parseLlmSceneResponse(raw: string) {
  // Remove code fences if present
  let cleaned = raw.replace(/```json|```/gi, "").trim();
  let parsed: any;
  try {
    parsed = JSON.parse(cleaned);
  } catch {
    // fallback: try to extract JSON object from text
    const match = cleaned.match(/\{[\s\S]*\}/);
    if (match) {
      try {
        parsed = JSON.parse(match[0]);
      } catch {
        parsed = {};
      }
    } else {
      parsed = {};
    }
  }
  // REPAIR dialog options before validation
  parsed = repairDialogOptions(parsed);
  // Validate and fill defaults
  return SceneResponseSchema.parse(parsed);
}

function repairDialogOptions(scene: any) {
  if (!scene || !Array.isArray(scene.dialog)) return scene;
  scene.dialog = scene.dialog.map((opt: any, idx: number) => {
    if (opt.skillCheck) {
      let repaired = false;
      let part = opt.skillCheck.part;
      let dc = opt.skillCheck.dc;
      if (typeof part !== "string") {
        part = "...";
        repaired = true;
      }
      if (typeof dc !== "number") {
        dc = -1;
        repaired = true;
      }
      if (repaired) {
        console.error("Repaired malformed skillCheck in dialog option", { idx, opt });
      }
      return { ...opt, skillCheck: { part, dc } };
    }
    return opt;
  });
  return scene;
}

