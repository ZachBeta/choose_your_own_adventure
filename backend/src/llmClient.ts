import fetch from 'node-fetch';
import { execSync } from 'child_process';

// Read OpenRouter API key from environment, fallback to Lean Vault
let OPENROUTER_API_KEY: string | undefined = process.env.OPENROUTER_API_KEY;
if (!OPENROUTER_API_KEY) {
  try {
    OPENROUTER_API_KEY = execSync('lean_vault get cyoa-gemma3-12b', { encoding: 'utf-8' }).trim() || undefined;
  } catch {
    OPENROUTER_API_KEY = undefined;
  }
}

import fs from 'fs';
import { logInfo } from './logger';


export class LLMClient {
  model: string;
  endpoint: string;
  headers: Record<string, string>;

  constructor() {
    if (OPENROUTER_API_KEY) {
      this.model = 'google/gemma-3-12b-it';
      this.endpoint = 'https://openrouter.ai/api/v1/chat/completions';
      this.headers = {
        'Authorization': `Bearer ${OPENROUTER_API_KEY}`,
        'Content-Type': 'application/json',
      };
    } else {
      this.model = 'gemma3:12b';
      this.endpoint = 'http://localhost:11434/api/generate';
      this.headers = { 'Content-Type': 'application/json' };
    }
  }

  async generate(prompt: string): Promise<string> {
    logInfo('>>> LLM PROMPT:', prompt);
    const body = OPENROUTER_API_KEY
      ? JSON.stringify({
          model: this.model,
          messages: [{ role: 'user', content: prompt }],
          temperature: 0
        })
      : JSON.stringify({
          model: this.model,
          prompt,
          stream: false,
          temperature: 0
        });
    let response: string;
    if (OPENROUTER_API_KEY) {
      const res = await fetch(this.endpoint, {
        method: 'POST',
        headers: this.headers,
        body,
      });
      const json = await res.json();
      response = json.choices[0]?.message?.content || '';
    } else {
      const res = await fetch(this.endpoint, {
        method: 'POST',
        headers: this.headers,
        body,
      });
      const json = await res.json();
      response = json.response || '';
    }

    logInfo('<<< LLM RESPONSE:', response);
    return response;
  }
}
