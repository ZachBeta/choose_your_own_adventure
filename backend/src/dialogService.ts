import { LLMClient } from './llmClient';

export class DialogService {
  llm: LLMClient;
  private playerStates: Map<string, { history: any[] }>;

  constructor({ llm }: { llm: LLMClient }) {
    this.llm = llm;
    this.playerStates = new Map();
  }

  async startScene({ playerId, sceneId }: { playerId: string; sceneId: string }) {
    // initialize player state
    if (!this.playerStates.has(playerId)) {
      this.playerStates.set(playerId, { history: [] });
    }

    const prompt = `You are an AI game master.\nGiven the following scene context, respond in strict JSON with these fields:\n- monologue: a short, vivid description of the scene\n- thoughtCabinet: an array of objects, each with 'part' (Joy, Fear, etc.) and 'text'\n- dialog: an array of dialog options, each with 'id', 'text', and optional 'skillCheck' (with 'part' and 'dc')\n\nScene context:\nPlayer has just entered scene "${sceneId}".\n\nRespond ONLY with valid JSON, no explanation or extra text. Example:\n{\n  "monologue": "...",\n  "thoughtCabinet": [\n    { "part": "Joy", "text": "..." }\n  ],\n  "dialog": [\n    { "id": "opt1", "text": "...", "skillCheck": { "part": "Joy", "dc": 10 } }\n  ]\n}`;
    const llmText = await this.llm.generate(prompt);
    // extract raw JSON string from any code fences or extra text
    const jsonMatch = llmText.match(/\{[\s\S]*\}/);
    const jsonString = jsonMatch ? jsonMatch[0] : llmText;
    let structured;
    try {
      console.log('LLM response:', llmText);
      structured = JSON.parse(jsonString);
    } catch (e) {
      throw new Error('LLM did not return valid JSON');
    }
    return {
      monologue: structured.monologue,
      thoughtCabinet: structured.thoughtCabinet,
      dialog: structured.dialog,
      llmLines: [jsonString]
    };
  }

  async chooseOption({ playerId, optionId }: { playerId: string; optionId: string }) {
    // retrieve and update history
    const state = this.playerStates.get(playerId);
    if (!state) throw new Error(`Unknown player ${playerId}`);
    const prompt = `You are an AI game master. The player history: ${JSON.stringify(
      state.history
    )}. The player chose option "${optionId}". Respond in strict JSON with these fields:
- monologue: a short, vivid description
- thoughtCabinet: array of { part, text }
- dialog: array of options { id, text, optional skillCheck { part, dc } }
- skillCheckResult: optional object with roll, part, dc, outcome
Respond ONLY with valid JSON.`;
    const llmText = await this.llm.generate(prompt);
    const jsonMatch = llmText.match(/\{[\s\S]*\}/);
    const jsonString = jsonMatch ? jsonMatch[0] : llmText;
    let structured;
    try {
      structured = JSON.parse(jsonString);
    } catch (e) {
      throw new Error('LLM did not return valid JSON for chooseOption');
    }
    // update history
    state.history.push({ optionId, result: structured.skillCheckResult });
    return {
      monologue: structured.monologue,
      thoughtCabinet: structured.thoughtCabinet,
      dialog: structured.dialog,
      skillCheckResult: structured.skillCheckResult ?? undefined,
      llmLines: [jsonString]
    };
  }

  getPlayerState({ playerId }: { playerId: string }) {
    const state = this.playerStates.get(playerId);
    if (!state) throw new Error(`Unknown player ${playerId}`);
    return { history: state.history };
  }
}
