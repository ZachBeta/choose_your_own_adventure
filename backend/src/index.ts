// Entry point for backend server

import express, { Request, Response } from 'express'
import cors from 'cors'
import { DialogService } from './dialogService'
import { LLMClient } from './llmClient'
import { logInfo, logError } from './logger';

const app = express()
app.use(cors(), express.json())

// Middleware to log all requests and responses
app.use((req, res, next) => {
  logInfo('API Request', {
    method: req.method,
    url: req.originalUrl,
    body: req.body,
    query: req.query,
    headers: req.headers
  });
  // Capture response
  const oldJson = res.json;
  res.json = function (data: any) {
    logInfo('API Response', {
      method: req.method,
      url: req.originalUrl,
      response: data
    });
    return oldJson.call(this, data);
  };
  next();
});

const dialog = new DialogService({ llm: new LLMClient() })

// POST /api/dialog  { playerId, sceneId? , optionId? }
app.post('/api/dialog', async (req: Request, res: Response) => {
  try {
    const { playerId, sceneId, optionId } = req.body
    let out = sceneId
      ? await dialog.startScene({ playerId, sceneId })
      : await dialog.chooseOption({ playerId, optionId })
    res.json(out)
  } catch (e:any) {
    logError('API Error', { endpoint: '/api/dialog', error: e.message });
    res.status(400).json({ error: e.message })
  }
})

// GET /api/player/state?playerId=â€¦
app.get('/api/player/state', (req: Request, res: Response) => {
  try {
    const { playerId } = req.query as any
    res.json(dialog.getPlayerState({ playerId }))
  } catch (e:any) {
    logError('API Error', { endpoint: '/api/player/state', error: e.message });
    res.status(404).json({ error: e.message })
  }
})

const PORT = process.env.PORT || 3000
app.listen(PORT, () => console.log(`listening on http://localhost:${PORT}`))
