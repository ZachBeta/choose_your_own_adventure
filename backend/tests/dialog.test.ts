import { describe, it, expect } from 'vitest';
import { DialogService } from '../src/dialogService';
import { LLMClient } from '../src/llmClient';

describe('DialogService', () => {
  it('returns a valid scene response when starting a scene', async () => {
    const dialogService = new DialogService({
      llm: new LLMClient()
    });
    const req = { playerId: 'player1', sceneId: 'scene_intro' };
    const res = await dialogService.startScene(req);

    expect(JSON.parse(res.llmLines[0])).toMatchSnapshot();
  }, 60000);

  it('handles a multi-turn happy path dialogue with memory', async () => {
    const dialogService = new DialogService({
      llm: new LLMClient()
    });
    const playerId = 'player_happy';
    // 1. Start the scene
    let res = await dialogService.startScene({ playerId, sceneId: 'scene_intro' });
    expect(JSON.parse(res.llmLines[0])).toMatchSnapshot('startScene');

    // 2. Choose a skill-check option
    const skillOption = res.dialog.find(o => o.skillCheck);
    expect(skillOption).toBeDefined();
    res = await dialogService.chooseOption({ playerId, optionId: skillOption!.id });
    expect(JSON.parse(res.llmLines[0])).toMatchSnapshot('chooseOption-skill');

    // 3. Choose a no-skill-check option, if present
    const noCheckOption = res.dialog.find(o => !o.skillCheck);
    if (noCheckOption) {
      res = await dialogService.chooseOption({ playerId, optionId: noCheckOption.id });
      expect(JSON.parse(res.llmLines[0])).toMatchSnapshot('chooseOption-no-skill');
    }

    // 4. Validate memory/history via getPlayerState
    if (typeof dialogService.getPlayerState === 'function') {
      const state = await dialogService.getPlayerState({ playerId });
      expect(state).toMatchSnapshot('playerState');
    }
  }, 60000);
});
