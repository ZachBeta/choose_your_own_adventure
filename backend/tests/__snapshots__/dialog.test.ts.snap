// Vitest Snapshot v1

exports[`DialogService > handles a multi-turn happy path dialogue with memory > chooseOption-no-skill 1`] = `
{
  "dialog": [
    {
      "id": "opt2",
      "text": "Follow the path deeper into the forest.",
    },
    {
      "id": "opt3",
      "skillCheck": {
        "dc": 12,
        "part": "Perception Check",
      },
      "text": "Examine the base of the nearest tree for clues.",
    },
    {
      "id": "opt4",
      "skillCheck": {
        "dc": 10,
        "part": "Perception Check",
      },
      "text": "Listen intently for the source of the rustling.",
    },
    {
      "id": "opt5",
      "text": "Turn back and retrace your steps.",
    },
  ],
  "monologue": "The air hangs thick and humid, smelling of damp earth and something vaguely floral, yet unsettling. Towering trees, their branches interwoven like skeletal fingers, blot out most of the sky. A narrow, overgrown path winds ahead, disappearing into the gloom. You hear the drip, drip, drip of water, and a rustling in the undergrowth that could be anything.",
  "skillCheckResult": null,
  "thoughtCabinet": [
    {
      "part": "Environment",
      "text": "The forest feels ancient and watchful. There's a sense of being observed, though you can't pinpoint the source.",
    },
    {
      "part": "Sound",
      "text": "The dripping water is rhythmic, almost hypnotic. The rustling is sporadic and difficult to place.",
    },
    {
      "part": "Feeling",
      "text": "A prickle of unease runs down your spine. This place doesn't feel welcoming.",
    },
  ],
}
`;

exports[`DialogService > handles a multi-turn happy path dialogue with memory > chooseOption-skill 1`] = `
{
  "dialog": [
    {
      "id": "opt1",
      "text": "Examine the locket closely.",
    },
    {
      "id": "opt2",
      "text": "Search the chamber for other exits or clues.",
    },
    {
      "id": "opt3",
      "text": "Touch the phosphorescent moss.",
    },
    {
      "id": "opt4",
      "text": "Leave the chamber and return to the previous hallway.",
    },
  ],
  "monologue": "The heavy oak door creaks open, revealing a circular chamber bathed in the eerie green glow of phosphorescent moss clinging to the damp stone walls. A single, ornate pedestal stands in the center, upon which rests a tarnished silver locket. The air hangs thick with the scent of mildew and something subtly… metallic.",
  "skillCheckResult": null,
  "thoughtCabinet": [
    {
      "part": "Environment",
      "text": "The moss is unusually bright, almost pulsating. It seems to be the only source of light.",
    },
    {
      "part": "Object",
      "text": "The locket is intricately carved with scenes of what appear to be celestial events – constellations and eclipses.",
    },
    {
      "part": "Scent",
      "text": "That metallic smell... could it be blood? Or something else entirely?",
    },
  ],
}
`;

exports[`DialogService > handles a multi-turn happy path dialogue with memory > playerState 1`] = `
{
  "history": [
    {
      "optionId": "opt1",
      "result": null,
    },
    {
      "optionId": "opt1",
      "result": null,
    },
  ],
}
`;

exports[`DialogService > handles a multi-turn happy path dialogue with memory > startScene 1`] = `
{
  "dialog": [
    {
      "id": "opt1",
      "skillCheck": {
        "dc": 12,
        "part": "Perception",
      },
      "text": "Approach the hooded figure.",
    },
    {
      "id": "opt2",
      "skillCheck": {
        "dc": 8,
        "part": "Insight",
      },
      "text": "Browse the stalls, starting with the potion vendor.",
    },
    {
      "id": "opt3",
      "skillCheck": {
        "dc": 10,
        "part": "Diplomacy",
      },
      "text": "Ask a nearby merchant for directions.",
    },
    {
      "id": "opt4",
      "skillCheck": {
        "dc": 15,
        "part": "Awareness",
      },
      "text": "Observe the crowd, trying to get a sense of the market's atmosphere.",
    },
  ],
  "monologue": "The air hangs thick and humid as you step into the Whispering Market. Stalls overflow with strange wares – shimmering fabrics, bubbling potions, and cages filled with chittering, unidentifiable creatures. A cacophony of voices rises around you, a blend of bartering, gossip, and the occasional shriek. The cobblestone path is slick with moisture, reflecting the flickering light of oil lamps that cast long, dancing shadows. A hooded figure watches you from across the square, their face obscured.",
  "thoughtCabinet": [
    {
      "part": "Curiosity",
      "text": "So many strange things... what should I look at first?",
    },
    {
      "part": "Caution",
      "text": "That figure... they seem to be observing me. Should I be worried?",
    },
    {
      "part": "Apprehension",
      "text": "The air feels heavy, almost oppressive. Something isn't quite right.",
    },
  ],
}
`;

exports[`DialogService > returns a valid scene response when starting a scene 1`] = `
{
  "dialog": [
    {
      "id": "opt1",
      "skillCheck": {
        "dc": 12,
        "part": "Perception",
      },
      "text": "Approach the hooded figure.",
    },
    {
      "id": "opt2",
      "skillCheck": {
        "dc": 8,
        "part": "Insight",
      },
      "text": "Browse the stalls, starting with the potion vendor.",
    },
    {
      "id": "opt3",
      "skillCheck": {
        "dc": 10,
        "part": "Diplomacy",
      },
      "text": "Ask a nearby merchant for directions.",
    },
    {
      "id": "opt4",
      "skillCheck": {
        "dc": 15,
        "part": "Awareness",
      },
      "text": "Observe the crowd, trying to get a sense of the market's atmosphere.",
    },
  ],
  "monologue": "The air hangs thick and humid as you step into the Whispering Market. Stalls overflow with strange wares – shimmering fabrics, bubbling potions, and cages filled with chittering, unidentifiable creatures. A cacophony of voices rises around you, a blend of bartering, gossip, and the occasional shriek. The cobblestone path is slick with moisture, reflecting the flickering light of oil lamps that cast long, dancing shadows. A hooded figure watches you from across the square, their face obscured.",
  "thoughtCabinet": [
    {
      "part": "Curiosity",
      "text": "So many strange things... what should I look at first?",
    },
    {
      "part": "Caution",
      "text": "That figure... they seem to be observing me. Should I be worried?",
    },
    {
      "part": "Apprehension",
      "text": "The air feels heavy, almost oppressive. Something isn't quite right.",
    },
  ],
}
`;
